### PR TITLE
iOS tv-casting-app v1.3 Commissioner-Generated passcode flow

### PR DESCRIPTION
**Pull request by pgregorr-amazon**
iOS tv-casting-app example app, implemented new Matter v1.3 Matter Casting User Directed Commissioning (UDC) feature Commissioner-Generated passcode commissioning flow.

**Change summary**

1.	Updated the iOS tv-casting-app example app CX with the CommissioneR-Generated passcode commissioning flow. Long pressing on a discovered casting player starts the new flow (MCDiscoveryExampleView.swift). Regular/short press starts the CommissioneE-generated passcodeflow as before this change.
2.	Updated examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleViewModel.swift to display an input dialog for the commissioner-generated passcode.
3.	New Objective-C files to enable the new flow: MCConnectionCallbacks, MCCommissionerDeclaration, MCIdentificationDeclarationOptions and MCTargetAppInfo.
4.	Updated MCCastingPlayer.h/mm with the new and modified native APIs:  verifyOrEstablishConnection(), continueConnecting() and stopConnecting().
5.	Updated examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCDataSource.h so that it can be updated with the user entered commissioner-generated passcode. Updated examples/tv-casting-app/darwin/TvCasting/TvCasting/MCInitializationExample.swift accordingly, corresponding to the MCDataSource change requiring the client to implement an update() function.
6.	Updated connectedhomeip/examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingApp.mm with the updateCommissionableDataProvider() function which the client needs to call to update the Data Provider with the user entered passcode.
7.	Updated examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer with the new and modified cpp API’s which call the common core/Linux implementations.
8.	Updated the shim layer verifyOrEstablishConnection() used by the PV mobile app in examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/compat-shim/CastingServerBridge.mm
9.	Updated the Swift View Model for demo interactions: MCApplicationBasicReadVendorIDExampleViewModel, MCContentLauncherLaunchURLExampleViewModel and MCMediaPlaybackSubscribeToCurrentStateExampleViewMode.
10. Added MCEndpointSelector.swift

**Testing**

Verified and tested locally with the Linux, Android and iOS tv-casting-app example mobile apps, and the Linux tv-app (CastingPlayer). With the iOS example tv-casting-app app running on the Xcode emulator, we are able to successfully commission the Linux tv-app running on another terminal on the same machine. Verified the CommissioneR-generated passcode flow by long pressing on a discovered CastingPlayer to start the commissioning flow. The original CommissioneE-generated passcode commissioning flow continues to function by regular/short pressing on the discovered CastingPlayer.

CommissioneR-Generated passcode commissioning flow - Continue Connecting

CommissioneR-Generated passcode commissioning flow - Cancel

CommissioneE-Generated passcode commissioning flow - old flow

[iOS tv-casting-app CommissioneE-generated pass.log](https://github.com/user-attachments/files/15880383/iOS.tv-casting-app.CommissioneE-generated.pass.log)
[iOS tv-casting-app CommissioneR-generated pass Cancel.log](https://github.com/user-attachments/files/15880385/iOS.tv-casting-app.CommissioneR-generated.pass.Cancel.log)
[iOS tv-casting-app CommissioneR-generated pass.log](https://github.com/user-attachments/files/15880386/iOS.tv-casting-app.CommissioneR-generated.pass.log)

### CODE REVIEW

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCEndpoint.h [hunk 1]
```diff
@@ -41,6 +41,9 @@
 - (BOOL)hasCluster:(MCEndpointClusterType)type;
 - (MCCluster * _Nullable)clusterForType:(MCEndpointClusterType)type;
 
+- (NSString * _Nonnull)description;
+- (void)logDetail;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCEndpoint.h:45]: Unlike description, logDetail is a non-standard function. Can't we just log the endpoint description by calling ChipLogDetail/Log directly, when we need to? It is one line either way. I'd rather not expose public APIs if they are not really needed.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+#import "MCTargetAppInfo.h"
+#import <Foundation/Foundation.h>
+
+#ifndef MCIdentificationDeclarationOptions_h
+#define MCIdentificationDeclarationOptions_h
+
+/**
+ * This class contains the optional parameters used in the IdentificationDeclaration Message, sent
+ * by the Commissionee to the Commissioner. The options specify information relating to the
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.h:25]: Add context. 

```suggestion
 * by the Commissionee (CastingApp) to the Commissioner (CastingPlayer). The options specify information relating to the
```

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * @brief Adds a TargetAppInfo to the IdentificationDeclarationOptions.java TargetAppInfos list,
+ *     up to a maximum of CHIP_DEVICE_CONFIG_UDC_MAX_TARGET_APPS.
+ */
+- (BOOL)addTargetAppInfo:(MCTargetAppInfo *)targetAppInfo NS_SWIFT_NAME(addTargetAppInfo(_:));
+- (NSArray<MCTargetAppInfo *> *)getTargetAppInfoList;
+
+- (NSString *)description;
+- (void)logDetail;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.h:49]: Same comment as MCEndpoint.h

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.mm [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#import "MCIdentificationDeclarationOptions.h"
+
+#import "core/IdentificationDeclarationOptions.h" // from tv-casting-common
+#import "core/Types.h" // from tv-casting-common
+#import "include/CHIPProjectAppConfig.h" // from tv-casting-common
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.mm:21]: This should not be needed. The build system should be including this overrides config automatically.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * @brief Feature: Target Content Application - An entry in the TargetAppList which contains
+ * a Target vendorId and an optional Target productId.
+ */
+@interface MCTargetAppInfo : NSObject
+
+/** Target Target Content Application Vendor ID, 0 means unspecified */
+@property (nonatomic) uint16_t vendorId;
+
+/** Target Target Content Application Product ID, 0 means unspecified */
+@property (nonatomic) uint16_t productId;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h:33]: These should not be exposed publicly. Instead add a MCTargetAppInfo\_Internal.h where you can put these. If required add getter methods for them to this public header file.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.mm [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+ * to exit the commissioning process.
+ */
+@property (nonatomic) BOOL cancelPasscode;
+/**
+ * Feature: Target Content Application - The set of content app Vendor IDs (and optionally,
+ * Product IDs) that can be used for authentication. Also, if TargetAppInfo is passed in,
+ * VerifyOrEstablishConnection() will force User Directed Commissioning, in case the desired
+ * TargetApp is not found in the on-device CastingStore.
+ */
+@property (nonatomic, strong) NSMutableArray<MCTargetAppInfo *> * targetAppInfos;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.mm:60]: You have a comment at the beginning saying "// Private properties", but these are all public. To actually make them "private" or internal-only, move them to the Internal header file.

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift [hunk 7]
```diff
@@ -30,66 +30,80 @@ class MCApplicationBasicReadVendorIDExampleViewModel: ObservableObject {
...
         // select the MCEndpoint on the MCCastingPlayer to invoke the command on
-        if let endpoint: MCEndpoint = castingPlayer.endpoints().filter({ $0.vendorId().intValue == sampleEndpointVid}).first
+        if let endpoint = castingPlayer.endpoints().filter({ $0.vendorId().intValue == sampleEndpointVid }).first {
+            selectedEndpoint = endpoint
+        // For the example Commissioner-Generated passcode commissioning flow, run demo interactions with the Endpoint with
+        // ID 1. For this flow, we commissioned with the Target Content Application with Vendor ID 1111. Since this target
+        // content application does not report its Endpoint's Vendor IDs, we find the desired endpoint based on the Endpoint
+        // ID. See connectedhomeip/examples/tv-app/tv-common/include/AppTv.h.
+        } else if let endpoint = castingPlayer.endpoints().filter({ $0.identifier().intValue == 1 }).first {
+            self.Log.info("MCApplicationBasicReadVendorIDExampleViewModel.read() No endpoint matching the sampleEndpointVid: \(String(describing: self.sampleEndpointVid)), but found endpoint with identifier: 1")
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift:45]: Like Android/Linux, we should pass in a flag to this ViewModel: useCommissionerGeneratedPasscode = true/false, to help us run the right endpoint selection logic.

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCDiscoveryExampleView.swift [hunk 8]
```diff
@@ -25,25 +25,30 @@ extension MCCastingPlayer : Identifiable {
 
 struct MCDiscoveryExampleView: View {
     @StateObject var viewModel = MCDiscoveryExampleViewModel()
+    @State private var selectedCastingPlayer: MCCastingPlayer?
+    // Enable navigating to the MCConnectionExampleView with or without the use
+    // Commissioner-Generated Passcode (CGP) flag.
+    @State private var navigateWithUseCGP = false
+    @State private var navigateWithoutUseCGP = false
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/TvCasting/TvCasting/MCDiscoveryExampleView.swift:32]: Not sure I understood the use of these two @State vars?

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer.mm [hunk 9]
```diff
@@ -41,53 +45,118 @@ + (NSInteger)kMinCommissioningWindowTimeoutSec
...
+            dispatch_queue_t clientQueue = [[MCCastingApp getSharedInstance] getClientQueue];
+            dispatch_async(clientQueue, ^{
+                if (connectionCallbacks.connectionCompleteCallback) {
+                    connectionCallbacks.connectionCompleteCallback(err == CHIP_NO_ERROR ? nil : [MCErrorUtils NSErrorFromChipError:err]);
+                } else {
+                    ChipLogError(AppServer, "MCCastingPlayer.verifyOrEstablishConnectionWithCallbacks() connectCallback(), client failed to set the connectionCompleteCallback() callback");
+                }
+            });
+        };
+        void (^commissionerDeclarationCallback)(const chip::Transport::PeerAddress & source, const chip::Protocols::UserDirectedCommissioning::CommissionerDeclaration cppCommissionerDeclaration) = ^(const chip::Transport::PeerAddress & source, const chip::Protocols::UserDirectedCommissioning::CommissionerDeclaration cppCommissionerDeclaration) {
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer.mm:94]: Since we now have multiple callbacks, can you please add a couple of lines of inline comments here and for the other callback to say what it is that we are setting up?

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer.mm [hunk 10]
```diff
@@ -41,53 +45,118 @@ + (NSInteger)kMinCommissioningWindowTimeoutSec
...
-        if (desiredEndpointFilter != nil) {
-            chip::Protocols::UserDirectedCommissioning::TargetAppInfo targetAppInfo;
-            targetAppInfo.vendorId = desiredEndpointFilter.vendorId;
-            targetAppInfo.productId = desiredEndpointFilter.productId;
-
-            CHIP_ERROR result = idOptions.addTargetAppInfo(targetAppInfo);
-            if (result != CHIP_NO_ERROR) {
-                ChipLogError(AppServer, "MCCastingPlayer.verifyOrEstablishConnectionWithCompletionBlock() failed to add targetAppInfo: %" CHIP_ERROR_FORMAT, result.Format());
-            }
+        matter::casting::core::IdentificationDeclarationOptions cppIdOptions;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer.mm:76]: This method implementation is getting pretty long. Can you move the setup of the cppIdOptions to a separate internal (private) method in MCCastingPlayer (do not expose publicly)?

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCConnectionCallbacks.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+#ifndef MCConnectionCallbacks_h
+#define MCConnectionCallbacks_h
+
+/** @brief A container class for User Directed Commissioning (UDC) callbacks. */
+@interface MCConnectionCallbacks : NSObject
+
+/**
+ * @param connectionCompleteCallback is required.
+ * @param commissionerDeclarationCallback is optional.
+ *
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCConnectionCallbacks.h:31]: Add a description of what the callbacks are here (you can move some of it from the verifyOrEstablishConnection docs) - similar to comments on previous Android PRs.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCommissionerDeclaration.h [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Represents the Commissioner Declaration message sent by a User Directed Commissioning server
+ * (CastingPlayer/Commissioner) to a UDC client (Casting Client/Commissionee).
+ */
+@interface MCCommissionerDeclaration : NSObject
+
+/** The allowed values for the ErrorCode field are the following */
+typedef NS_ENUM(NSInteger, CdError) {
+    kNoError = 0,
+    kCommissionableDiscoveryFailed = 1,
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCommissionerDeclaration.h:31]: Is it possible to add the string representation of these CdErrors to this enum instead of needing the separate stringForErrorCode method?

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleView.swift [hunk 13]
```diff
@@ -17,19 +17,32 @@
...
     }
     
     var body: some View {
         VStack(alignment: .leading) {
-            Text("Verifying or Establishing Connection to Casting Player: \(self.selectedCastingPlayer!.deviceName())").padding()
+            if self.useCommissionerGeneratedPasscode {
+                if selectedCastingPlayer?.supportsCommissionerGeneratedPasscode() == true {
+                    Text("Verifying or Establishing Connection, using Casting Player/Commissioner-Generated passcode, with Casting Player: \(self.selectedCastingPlayer!.deviceName())\n\nEnter the passcode displayed on the Casting Player when prompted.").padding()
+                } else {
+                    Text("Sorry, the selected Casting Player does not support Casting Player/Commissioner-Generated passcode commissioning. \n\nCasting Player selected: \(self.selectedCastingPlayer!.deviceName()) \n\nRoute back to select a different Casting Player.").padding()
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleView.swift:41]:
```suggestion
                    Text("\(self.selectedCastingPlayer!.deviceName()) does not support Casting Player/Commissioner-Generated passcode commissioning. \n\nSelect a different Casting Player.").padding()
```

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleView.swift [hunk 14]
```diff
@@ -56,13 +69,23 @@ struct MCConnectionExampleView: View {
         .navigationTitle("Connecting...")
         .frame(minWidth: 0, maxWidth: .infinity, minHeight: 0, maxHeight: .infinity, alignment: .top)
         .onAppear(perform: {
-            viewModel.connect(selectedCastingPlayer: self.selectedCastingPlayer)
+            if self.useCommissionerGeneratedPasscode {
+                if selectedCastingPlayer?.supportsCommissionerGeneratedPasscode() == true {
+                    self.Log.info("MCConnectionExampleView calling MCConnectionExampleViewModel.connect() with useCommissionerGeneratedPasscode: \(String(describing: self.useCommissionerGeneratedPasscode))")
+                    viewModel.connect(selectedCastingPlayer: self.selectedCastingPlayer, useCommissionerGeneratedPasscode: self.useCommissionerGeneratedPasscode)
+                } else {
+                    self.Log.error("MCConnectionExampleView Sorry, the selected Casting Player does not support Commissioner-Generated passcode commissioning. Casting Player: \(self.selectedCastingPlayer!.deviceName()), Route back to try again.")
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleView.swift:77]:
```suggestion
                    self.Log.error("MCConnectionExampleView \(self.selectedCastingPlayer!.deviceName()) does not support Casting Player/Commissioner-Generated passcode commissioning. \n\nSelect a different Casting Player.")
```

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleViewModel.swift [hunk 15]
```diff
@@ -15,39 +15,179 @@
  *    limitations under the License.
  */
 
-
 import Foundation
 import os.log
+import UIKit
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleViewModel.swift:20]: Why is this needed?

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/compat-shim/CastingServerBridge.mm [hunk 16]
```diff
@@ -226,11 +228,20 @@ - (void)sendUserDirectedCommissioningRequest:(DiscoveredNodeData * _Nonnull)comm
 {
     ChipLogProgress(AppServer, "CastingServerBridge().sendUserDirectedCommissioningRequest() called with desiredContentAppVendorId: %d", desiredContentAppVendorId);
 
-    MCEndpointFilter * filter = [MCEndpointFilter new];
-    filter.vendorId = desiredContentAppVendorId;
+    MCIdentificationDeclarationOptions * identificationDeclarationOptions = [[MCIdentificationDeclarationOptions alloc] init];
+    MCTargetAppInfo * targetAppInfo = [[MCTargetAppInfo alloc] init];
+    targetAppInfo.vendorId = desiredContentAppVendorId;
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/compat-shim/CastingServerBridge.mm:233]: Pass this in via constructor.

**Review by sharadb-amazon — CHANGES_REQUESTED** [review]

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingApp.h [hunk 17]
```diff
@@ -42,6 +42,16 @@
...
 
+/**
+ * @brief Updates the MCCommissionableDataProvider stored in this CastingApp with the new CommissionableData
+ * to be used for the next commissioning session. Calling this function is mandatory for the
+ * Casting Player/Commissioner-Generated passcode commissioning flow, since the commissioning session's PAKE
+ * verifier needs to be updated with the user entered passcode.
+ *
+ * @param newCommissionableData the new MCCommissionableData to be used for the next commissioning session.
+ */
+- (NSError * _Nullable)updateCommissionableDataProvider:(MCCommissionableData * _Nonnull)newCommissionableData;
```
[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingApp.h:53]: Ok, we can do that for the sake of consistency. updateCommissionableDataProvider() will be called from MCCastingPlayer.continueConnecting(). We'll store the MCAppParametersDataSource in an singleton instance of the MCInitializationExample so that can be accessed and updated by the client form MCConnectionExampleViewModel.swift (prior to calling MCCastingPlayer.continueConnecting()).

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer.mm (hunk 10 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCastingPlayer.mm:76]: Adding setupCppIdOptions method defined as a private method within the @interface section. In Objective-C, methods defined within a class's @implementation section are effectively private if they are not declared in the public header file.
However "pretty long" is arbitrary, we should define a concrete standard.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCommissionerDeclaration.h (hunk 12 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCCommissionerDeclaration.h:31]: I Googled around, Objective-C doesn't support adding metadata directly to enum values.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCConnectionCallbacks.h (hunk 11 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCConnectionCallbacks.h:31]: This is the same description as from ConnectionCallbacks.java in the Android PR. Adding some more detail.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCEndpoint.h (hunk 1 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCEndpoint.h:45]: We make logDetail available in a number of functions across iOS, Linux and Android versions of Endpoint and other classes. For consistency I added it here as well.  Also [endpoint logDetail]; is more compact than ChipLogDetail(AppServer, "MCEndpoint Details: %@", [self description]);

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.h (hunk 2 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.h:25]: Ok, updating the Java version too.

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.mm (hunk 6 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCIdentificationDeclarationOptions.mm:60]: Something like this "identificationDeclarationOptions.commissionerPasscode = false" will lead to a compile error: Value of type 'MCIdentificationDeclarationOptions' has no member 'commissionerPasscode'

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h (hunk 5 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h:33]: What is difference between the Android and Linux implementations that requires these to not be exposed publicly here? On Android and Linux this object is a simple class/struct.

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleViewModel.swift (hunk 15 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/TvCasting/TvCasting/MCConnectionExampleViewModel.swift:20]: For the new UX functionality to display the PasscodeInputDialog

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCDiscoveryExampleView.swift (hunk 8 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/TvCasting/TvCasting/MCDiscoveryExampleView.swift:32]: The @State variables navigateWithUseCGP and navigateWithoutUseCGP are used to control navigation to MCConnectionExampleView with different configurations based on the type of user interaction (tap or long press). They enable the view to distinguish between two different ways of connecting to a CastingPlayer. We can simplify the code by using a single variable and distinguishing between the two gestures within the navigation link's isActive binding.

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift (hunk 7 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift:45]: There is an advantage in falling back since the sample vendorID 65521 is actually available here and is our preferred content app vID for sample interactions. Something might have recently changed on the TV-app side recently; I'll file a JIRA to investigate and possibly change Android/Linux to use 65521 for demo interactions as well. We also avoid passing that flag useCommissionerGeneratedPasscode, which can be confusing since after commissioning it should not matter how we commissioned the Casting Player. 

MCApplicationBasicReadVendorIDExampleViewModel.read()
MCCastingPlayer logAllEndpoints():
MCCastingPlayer MCEndpoint details: <MCEndpoint: identifier=1, vendorId=0, productId=0>
MCCastingPlayer MCEndpoint details: <MCEndpoint: identifier=2, vendorId=0, productId=0>
MCCastingPlayer MCEndpoint details: <MCEndpoint: identifier=4, vendorId=65521, productId=32768>
MCApplicationBasicReadVendorIDExampleViewModel.read() selected endpoint: <MCEndpoint: identifier=4, vendorId=65521, productId=32768>

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCDataSource.h [hunk 18]
```diff
@@ -29,6 +29,16 @@
...
 
+/**
+ * @brief Updates the MCCommissionableData object stored inside MCAppParametersDataSource, which was used to
+ * initialize the MCCastingApp. This function needs to be implemented by the client in use cases where the
+ * MCCommissionableData needs to be updated post-initialization. For example, when the Commissioner-Generated
+ * Passcode feature is used.
+ *
+ * @param newCommissionableData The new MCCommissionableData object to be used for the next commissioning session.
+ */
+- (void)update:(MCCommissionableData * _Nonnull)newCommissionableData;
```
[the remark this answers is not in the document]
Comment (sharadb-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCDataSource.h:40]: Why aren't we just re-using castingAppDidReceiveRequestForCommissionableData that is already defined below?

[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCDataSource.h:40]: update -  Used to change the commissionableData. It is explicitly called by the client to update the data.
castingAppDidReceiveRequestForCommissionableData - Used to retrieve the current commissionableData. It is called by the casting app when it needs the data.

[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCDataSource.h:40]: removing  it from the MCDataSource protocol only

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+/** Initialize with vendorId and productId */
+- (instancetype _Nonnull)initWithVendorId:(uint16_t)vendorId productId:(uint16_t)productId NS_DESIGNATED_INITIALIZER;
+
+/** Vendor ID getter */
+- (uint16_t)getVendorId;
+
+/** Vendor ID setter */
+- (void)setVendorId:(uint16_t)vendorId;
+
```
[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h:44]: removing

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift [hunk 20]
```diff
@@ -30,66 +30,80 @@ class MCApplicationBasicReadVendorIDExampleViewModel: ObservableObject {
 
     func read(castingPlayer: MCCastingPlayer)
     {
+        self.Log.info("MCApplicationBasicReadVendorIDExampleViewModel.read()")
+        castingPlayer.logAllEndpoints()
```
[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift:34]: OK.. when you update the APIs.md docs, just omit these logging lines in the inline examples. The APIs.md docs are supposed to just have minimal excerpts to help explain how to use the APIs.

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/TvCastingApp.swift [hunk 21]
```diff
@@ -33,7 +33,8 @@ struct TvCastingApp: App {
                     {
                         self.Log.info("CHIP_CASTING_SIMPLIFIED = 1")
                         
-                        let err: Error? = MCInitializationExample().initialize()
+                        let initializationExample = MCInitializationExample.shared
+                        let err: Error? = initializationExample.initialize()
```
[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/TvCasting/TvCasting/TvCastingApp.swift:37]: ok, wasn't sure if this was going to run with adding the "shared" but it is:
let err: Error? = MCInitializationExample.shared.initialize()

File: examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift (hunk 7 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/TvCasting/TvCasting/MCApplicationBasicReadVendorIDExampleViewModel.swift:45]: Adding MCEndpointSelector.swift

File: examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+- (instancetype)init;
+
+/** Initialize with vendorId, productId defaults to 0 */
+- (instancetype _Nonnull)initWithVendorId:(uint16_t)vendorId;
+
+/** Initialize with vendorId and productId */
+- (instancetype _Nonnull)initWithVendorId:(uint16_t)vendorId productId:(uint16_t)productId NS_DESIGNATED_INITIALIZER;
+
+/** Vendor ID getter */
+- (uint16_t)getVendorId;
```
[the remark this answers is not in the document]
Comment (pgregorr-amazon) [reply examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCTargetAppInfo.h:40]: Ok

**Review by sharadb-amazon — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

**Review by hawk248 — APPROVED** [review]
